### PR TITLE
[Ripple/Near] new transaction API

### LINKED ISSUE #1402: Transaction API v2
https://github.com/trustwallet/golibs/issues/31

### PR DESCRIPTION
**Pull request by hewigovens**
Fixes #1402 and fixes #1409

Changes: 

1. Save v2 compatible transfer, token transfer txs to db
2. Add new endpoint to fetch txs by account
3. Near block parsing

### CODE REVIEW

File: api/endpoint/transaction.go [hunk 1]
```diff
@@ -95,6 +96,65 @@ func GetTransactionsHistory(c *gin.Context, txAPI blockatlas.TxAPI, tokenTxAPI b
...
+		c.AbortWithStatusJSON(http.StatusBadRequest, errorResponse(blockatlas.ErrInvalidAddr))
+		return
+	}
+	token := c.Query("token")
+
+	txs, err := txsAPI.GetTransactionsByAccount(account, token, database)
+
+	if err != nil {
+		switch err {
+		case blockatlas.ErrInvalidAddr:
```
Comment (vikmeup) [inline api/endpoint/transaction.go:111]: those error handlers not relevant for this case

File: api/endpoint/transaction.go [hunk 2]
```diff
@@ -95,6 +96,65 @@ func GetTransactionsHistory(c *gin.Context, txAPI blockatlas.TxAPI, tokenTxAPI b
...
+			return
+		default:
+			c.AbortWithStatusJSON(
+				http.StatusInternalServerError,
+				errorResponse(err),
+			)
+			return
+		}
+	}
+	var (
```
Comment (vikmeup) [inline api/endpoint/transaction.go:137]: transactions should be already sorted by the db query

File: api/endpoint/transaction.go [hunk 3]
```diff
@@ -95,6 +96,65 @@ func GetTransactionsHistory(c *gin.Context, txAPI blockatlas.TxAPI, tokenTxAPI b
...
+		tx.Direction = tx.GetTransactionDirection(account)
+		page = append(page, tx)
+	}
+
+	page = page.FilterTransactionsByMemo()
+	if token != "" {
+		page = page.FilterTransactionsByToken(token)
+	}
+
+	if len(page) > types.TxPerPage {
```
Comment (vikmeup) [inline api/endpoint/transaction.go:151]: same here, db should limit this

File: db/models/tracker.go [hunk 4]
```diff
@@ -3,9 +3,10 @@ package models
...
-	Coin      string `gorm:"primary_key:true; type:varchar(64)"`
-	Priority  string `gorm:"default:normal"`
-	Height    int64
-	Enabled   bool `gorm:"default:true" sql:"index"`
+	UpdatedAt     time.Time
+	Coin          string `gorm:"primary_key:true; type:varchar(64)"`
+	Priority      string `gorm:"default:normal"`
+	Height        int64  `gorm:"default:0"`
+	Enabled       bool   `gorm:"default:true" sql:"index"`
+	IndexedHeight int64  `gorm:"default:0"`
```
Comment (vikmeup) [inline db/models/tracker.go:11]:
```suggestion
	TransactionIndexHeight int64  `gorm:"default:0"`
```

File: db/transaction.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	"github.com/trustwallet/blockatlas/db/models"
+	"gorm.io/gorm/clause"
+)
+
+func (i *Instance) CreateTransactions(txs []models.Transaction) error {
+	return i.Gorm.Clauses(clause.OnConflict{DoNothing: true}).Create(&txs).Error
+}
+
+func (i *Instance) GetTransactionsByAccount(account string, coin uint) (txs []models.Transaction, err error) {
+	err = i.Gorm.Where("coin = ? AND (\"from\" = ? OR \"to\" = ?)", coin, account, account).
```
Comment (vikmeup) [inline db/transaction.go:13]: https://gorm.io/docs/query.html#Or-Conditions

File: db/transaction.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+		Order("date DESC").
+		Limit(25).
```
Comment (vikmeup) [inline db/transaction.go:15]: limit should be passed from the rest api, and set max limit to 25

File: services/tokenindexer/indexer.go [hunk 7]
```diff
@@ -138,3 +150,40 @@ func assetsMap(txs types.Txs) map[string][]string {
...
+		return nil
+	}
+	return database.CreateTransactions(txs)
+}
+
+func transactionsFrom(txs []types.Tx) []models.Transaction {
+	results := make([]models.Transaction, 0)
+	for _, tx := range txs {
+		metadata, err := json.Marshal(tx.Meta)
+		if err != nil {
```
Comment (vikmeup) [inline services/tokenindexer/indexer.go:165]: maybe create normalize() (models.Transaction, error) instead?

**Comment by hewigovens** [comment]
https://github.com/trustwallet/blockatlas/pull/1409 merged into this branch

File: platform/near/transaction.go [hunk 8]
```diff
@@ -13,6 +15,14 @@ func (p *Platform) GetTxsByAddress(address string) (types.TxPage, error) {
 	return normalized, nil
 }
 
+func (p *Platform) GetTransactionsByAccount(account, token string, limit int, database *db.Instance) (page types.TxPage, err error) {
```
Comment (vikmeup) [inline platform/near/transaction.go:18]: can we get rid of `types.TxPage` type? it's confusing. We should easier use `types.Tx{}` or `types.Txs`

Comment (hewigovens) [reply platform/near/transaction.go:18]: Ok, will remove it from golibs

File: pkg/blockatlas/platform.go [hunk 9]
```diff
@@ -1,6 +1,7 @@
 package blockatlas
 
 import (
+	"github.com/trustwallet/blockatlas/db"
```
Comment (vikmeup) [inline pkg/blockatlas/platform.go:4]: in `RunSubscriber` for `types.DeleteSubscription` we need to remove all transactions for `to`, `from` when specified

**Pull request closed without merging**